### PR TITLE
Log fatal errors to stderr, add exceptions to 'marked' errors

### PR DESCRIPTION
**Pull request by raboof**
Refs #23107
After logging to stderr it seems reasonable to include the exception
in the logging though the logging subsystem

### CODE REVIEW

File: akka-actor/src/main/scala/akka/actor/ActorSystem.scala [hunk 1]
```diff
@@ -666,24 +666,34 @@ private[akka] class ActorSystemImpl(
...
               terminate()
             }
         }
       }
+
+      @inline
+      private def logFatalError(message: String, cause: Throwable, thread: Thread): Unit = {
+        // First log to stderr as this has the best chance to get through in an 'emergency panic' situation:
+        import System.err
+        err.print("Uncaught error from thread [")
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/actor/ActorSystem.scala]: Maybe add `cause.getMessage` as well here.

File: akka-actor/src/main/scala/akka/actor/ActorSystem.scala [hunk 2]
```diff
@@ -666,24 +666,34 @@ private[akka] class ActorSystemImpl(
...
-                cause.printStackTrace(System.err)
-                System.err.flush()
+                logFatalError("shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for", cause, thread)
               } finally {
                 System.exit(-1)
               }
             } else {
-              markerLogging.error(LogMarker.Security, cause, "Uncaught fatal error from thread [{}] shutting down ActorSystem [{}]", thread.getName, name)
+              logFatalError("shutting down", cause, thread)
               terminate()
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/actor/ActorSystem.scala]: Run `terminate` in the finally block as well?

File: akka-actor-tests/src/test/scala/akka/event/MarkerLoggingSpec.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      system.eventStream.publish(TestEvent.Mute(EventFilter.info()))
+      markerLogging.info(LogMarker.Security, "This is a security problem")
+
+      val info = expectMsgType[Info3]
+      info.message should be("This is a security problem")
+      info.marker.name should be("SECURITY")
+    }
+
+    "allow cause exceptions in error messages" in {
+      class MyException(message: String) extends Exception(message)
```
Comment (jrudolph) [inline akka-actor-tests/src/test/scala/akka/event/MarkerLoggingSpec.scala]: Woops, great find. This fix could be in an extra commit.

**Review by jrudolph — APPROVED** [review]
Mostly LGTM, small nitpicks.

File: akka-actor/src/main/scala/akka/actor/ActorSystem.scala (hunk 2 above)
Comment (raboof) [inline akka-actor/src/main/scala/akka/actor/ActorSystem.scala]: `logFatalError` is pretty simple, if that crashes I'm not sure there's much of a chance `terminate()` will do anything useful... might make the code a bit more symmetrical though :)

Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/ActorSystem.scala]: +1 to `finally`

File: akka-actor/src/main/scala/akka/event/Logging.scala [hunk 4]
```diff
@@ -1456,31 +1456,31 @@ class MarkerLoggingAdapter(
...
+    if (isErrorEnabled) bus.publish(Error(cause, logSource, logClass, format1(template, arg1), mdc, marker))
 
   /**
    * Message template with 2 replacement arguments.
    * The marker argument can be picked up by various logging frameworks such as slf4j to mark this log statement as "special".
    * @see [[LoggingAdapter]]
    */
   def error(marker: LogMarker, cause: Throwable, template: String, arg1: Any, arg2: Any): Unit =
-    if (isErrorEnabled) bus.publish(Error(logSource, logClass, format(template, arg1, arg2), mdc, marker))
+    if (isErrorEnabled) bus.publish(Error(cause, logSource, logClass, format(template, arg1, arg2), mdc, marker))
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: wow!

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-actor/src/main/scala/akka/actor/ActorSystem.scala [hunk 5]
```diff
@@ -671,8 +671,11 @@ private[akka] class ActorSystemImpl(
                 System.exit(-1)
               }
             } else {
-              logFatalError("shutting down", cause, thread)
-              terminate()
+              try {
```
Comment (jrudolph) [inline akka-actor/src/main/scala/akka/actor/ActorSystem.scala]: Nitpick: no braces needed ;)

Comment (raboof) [inline akka-actor/src/main/scala/akka/actor/ActorSystem.scala]: that is actually quite a lot nicer :+1:

**Comment by jrudolph** [comment]
@johanandren just pointed out that #22324 already tracks the MarkerLoggingAdapter issue.

**Pull request merged**